packages/repository-tests/src/crud/relations/acceptance/has-many-inclusion-resolver.relation.acceptance.ts

```diff
@@ -0,0 +1,273 @@
+// Copyright IBM Corp. 2019. All Rights Reserved.
+// Node module: @loopback/repository-tests
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {expect, skipIf, toJSON} from '@loopback/testlab';
+import {Suite} from 'mocha';
+import {
+  CrudFeatures,
+  CrudRepositoryCtor,
+  CrudTestContext,
+  DataSourceOptions,
+} from '../../..';
+import {
+  deleteAllModelsInDefaultDataSource,
+  MixedIdType,
+  withCrudCtx,
+} from '../../../helpers.repository-tests';
+import {
+  Customer,
+  CustomerRepository,
+  Order,
+  OrderRepository,
+} from '../fixtures/models';
+import {givenBoundCrudRepositories} from '../helpers';
+
+export function hasManyRelationAcceptance(
+  dataSourceOptions: DataSourceOptions,
+  repositoryClass: CrudRepositoryCtor,
+  features: CrudFeatures,
+) {
+  skipIf<[(this: Suite) => void], void>(
+    !features.supportsInclusionResolvers,
+    describe,
+    'retrieve models including relations',
+    () => {
+      describe('HasMany inclusion resolvers - acceptance', () => {
+        before(deleteAllModelsInDefaultDataSource);
+        let customerRepo: CustomerRepository;
+        let orderRepo: OrderRepository;
+        let existingCustomerId: MixedIdType;
+
+        before(
+          withCrudCtx(async function setupRepository(ctx: CrudTestContext) {
+            // when running the test suite on MongoDB, we don't really need to setup
+            // this config for mongo connector to pass the test.
+            // however real-world applications might have such config for MongoDB
+            // setting it up to check if it works fine as well
+            Order.definition.properties.customerId.type = features.idType;
+            Order.definition.properties.customerId.mongodb = {
+              dataType: 'ObjectID',
+            };
+            // this helper should create the inclusion resolvers for us
+            ({customerRepo, orderRepo} = givenBoundCrudRepositories(
+              ctx.dataSource,
+              repositoryClass,
+            ));
+            // inclusionResolvers should be defined. And resolver for each
+            // relation should be created by the hasManyFactory at this point.
+            expect(customerRepo.inclusionResolvers).to.not.be.undefined();
+            expect(orderRepo.inclusionResolvers).to.not.be.undefined();
+            expect(customerRepo.orders.inclusionResolver).to.not.be.undefined();
+            expect(
+              customerRepo.customers.inclusionResolver,
+            ).to.not.be.undefined();
+            // inclusionResolvers shouldn't setup yet at this point
+            expect(customerRepo.inclusionResolvers).to.deepEqual(new Map());
```

Comment:
I feel some of the assertions are slightly redundant here. if `customerRepo.inclusionResolvers` is deep-equal to a Map, then surely it must not be undefined?

In general, I find the assertion `.not.be.undefined()` as not very helpful, because it accepts any value, including strings, numbers or empty object. I prefer positive assertions, even if the check is very broad e.g. `.to.be.Function()`.

I am proposing to rewrite the block on lines 60-67 as follows:

```ts
            expect(customerRepo.inclusionResolvers).to.deepEqual(new Map());
            expect(orderRepo.inclusionResolvers).to.deepEqual(new Map());
            expect(customerRepo.orders.inclusionResolver).to.be.Function();
            expect(customerRepo.customers.inclusionResolver).to.be.Function();
```

Reply:
I will remove `expect(customerRepo.inclusionResolvers).to.deepEqual(new Map());` since we decide to move registering resolver part to the helper function `givenBoundCrudRepositories`.